README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->
[![pub](https://img.shields.io/pub/v/simple_item_selector.svg)](https://pub.dev/packages/simple_item_selector)
[![license: MIT](https://img.shields.io/badge/license-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

This package contains ItemSelector(), a simple scrollable item selector widget. when developers have a list of items or widgets, then they can use ItemSelector() widget to build a simple and beautiful item selector in Flutter App. 

## Preview

| Horizontal | Vertical |
|:---:|:---:|
| ![horizontal](https://github.com/hosseinvejdani/simple_item_selector/assets/18458798/6f942a7f-ed78-4d3b-ab5b-24098bb8f59c) | ![vertical](https://github.com/hosseinvejdani/simple_item_selector/assets/18458798/5da7e941-60e9-4b41-8337-3eb913eea86f) |


## Features
Developers can pass any arbitrary list of widgets as items in horizontal or vertical axis directions. And Users can iteract with items and they can select one of them with a simple tap. Developer have access to the index of the selected item with onSelected attribute.

## Getting started

To use this package, add simple_item_selector as a dependency in your pubspec.yaml file. Then import the package like this:
```
import 'package:simple_item_selector/simple_item_selector.dart';
```

## Usage
This is Simple example:

```dart
ItemSelector(
  direction: Direction.horizontal,
  activeBackgroundColor: Colors.amberAccent,
  inactiveBackgroundColor: Colors.grey[300],
  itemMargin: const EdgeInsets.all(2.0),
  itemPadding: const EdgeInsets.all(15.0),
  itemBorderRadius: const BorderRadius.all(Radius.circular(5.0)),
  itemsCount: 4, // should be <= items.length
  items: const [Text('a1'), Text('a2'), Text('a3'), Text('a4')], // any arbitrary widget list
  onSelected: (index) {
    // you can access selected item index here!
    print(index);
  },
)
```


CHANGELOG.md
## [0.0.5] 
* minor changes.
  
## [0.0.4] 
* added new gif's for preview images.

## [0.0.3] 
* modify preview images.
  
## [0.0.2] 
* add preview images.
  
## [0.0.1] 
* created the packge for first time.



lib/simple_item_selector.dart
// ignore_for_file: must_be_immutable

library simple_item_selector;

import 'package:flutter/material.dart';

enum Direction { vertical, horizontal }

typedef OnSelected = void Function(int? index);

class ItemSelector extends StatefulWidget {
  int? initIndex;
  final int itemsCount;
  final List<Widget?> items;
  final OnSelected onSelected;
  final Color? activeBackgroundColor;
  final Color? inactiveBackgroundColor;
  final EdgeInsetsGeometry? itemPadding;
  final EdgeInsetsGeometry? itemMargin;
  final Direction? direction;
  final BorderRadius? itemBorderRadius;
  final Border? itemBorder;
  //
  ItemSelector({
    Key? key,
    required this.itemsCount,
    required this.items,
    required this.onSelected,
    this.initIndex = 0,
    this.activeBackgroundColor,
    this.inactiveBackgroundColor,
    this.itemPadding,
    this.itemMargin,
    this.direction = Direction.horizontal,
    this.itemBorderRadius,
    this.itemBorder,
  }) : super(key: key);

  @override
  State<ItemSelector> createState() => _ItemSelectorState();
}

class _ItemSelectorState extends State<ItemSelector> {
  //
  int? _currentIndex;
  Color? activeBackgroundColor;
  Color? inactiveBackgroundColor;
  EdgeInsetsGeometry? itemPadding;
  EdgeInsetsGeometry? itemMargin;
  //
  @override
  void initState() {
    _currentIndex = widget.initIndex;
    super.initState();
  }

  void _handleOnTap(int index) async {
    setState(() => _currentIndex = index);
    widget.onSelected(index);
  }

  //
  @override
  Widget build(BuildContext context) {
    /// Assigns active background color to default primary theme color if it's null/not provided.
    activeBackgroundColor = widget.activeBackgroundColor ?? Theme.of(context).colorScheme.primary;

    /// Assigns inactive background color to default disabled theme color if it's null/not provided.
    inactiveBackgroundColor = widget.inactiveBackgroundColor ?? Theme.of(context).disabledColor;

    /// Assings padding if its default value is null!
    itemPadding = widget.itemPadding ?? const EdgeInsets.all(8.0);

    /// Assings margin if its default value is null!
    itemMargin = widget.itemMargin ?? const EdgeInsets.all(0.0);

    // ================================
    return widget.direction == Direction.horizontal
        ? Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              for (int index = 0; index < widget.itemsCount; index++) singleItem(index),
            ],
          )
        : Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              for (int index = 0; index < widget.itemsCount; index++) singleItem(index),
            ],
          );
  }

  GestureDetector singleItem(int index) {
    return GestureDetector(
      onTap: () => _handleOnTap(index),
      child: Container(
        margin: itemMargin,
        padding: itemPadding,
        decoration: BoxDecoration(
          borderRadius: widget.itemBorderRadius,
          border: widget.itemBorder,
          color: _currentIndex == index ? activeBackgroundColor : inactiveBackgroundColor,
        ),
        child: widget.items[index],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:simple_item_selector/simple_item_selector.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const Scaffold(body: Home()),
    );
  }
}

class Home extends StatelessWidget {
  const Home({super.key});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ItemSelector(
        direction: Direction.vertical,
        activeBackgroundColor: Colors.amberAccent,
        inactiveBackgroundColor: Colors.grey[300],
        itemMargin: const EdgeInsets.all(2.0),
        itemPadding: const EdgeInsets.all(15.0),
        itemBorderRadius: const BorderRadius.all(Radius.circular(5.0)),
        itemsCount: 4, // should be <= items.length
        items: const [Text('a1'), Text('a2'), Text('a3'), Text('a4')], // any arbitrary widget list
        onSelected: (index) {
          // you can access selected item index here!
          print(index);
        },
      ),
    );
  }
}
